File: bindings/dart/test/lua_runner_test.dart
import 'package:psr_database/psr_database.dart';
import 'package:test/test.dart';
import 'package:path/path.dart' as path;

void main() {
  // Path to central tests folder
  final testsPath = path.join(
    path.current,
    '..',
    '..',
    'tests',
  );

  group('LuaRunner Create Element', () {
    test('creates element from Lua script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          lua.run('''
            db:create_element("Configuration", { label = "Test Config" })
            db:create_element("Collection", { label = "Item 1", some_integer = 42 })
          ''');

          final labels = db.readScalarStrings('Collection', 'label');
          expect(labels.length, equals(1));
          expect(labels[0], equals('Item 1'));

          final integers = db.readScalarIntegers('Collection', 'some_integer');
          expect(integers.length, equals(1));
          expect(integers[0], equals(42));
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Read from Lua', () {
    test('reads scalar strings in Lua script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        db.createElement('Configuration', {'label': 'Config'});
        db.createElement('Collection', {'label': 'Item 1', 'some_integer': 10});
        db.createElement('Collection', {'label': 'Item 2', 'some_integer': 20});

        final lua = LuaRunner(db);
        try {
          lua.run('''
            local labels = db:read_scalar_strings("Collection", "label")
            assert(#labels == 2, "Expected 2 labels")
            assert(labels[1] == "Item 1", "First label mismatch")
            assert(labels[2] == "Item 2", "Second label mismatch")
          ''');
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Script Error', () {
    test('throws LuaException for syntax error', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          expect(
            () => lua.run('invalid syntax !!!'),
            throwsA(isA<LuaException>()),
          );
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Reuse Runner', () {
    test('runs multiple scripts with same runner', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          lua.run('db:create_element("Configuration", { label = "Config" })');
          lua.run('db:create_element("Collection", { label = "Item 1" })');
          lua.run('db:create_element("Collection", { label = "Item 2" })');

          final labels = db.readScalarStrings('Collection', 'label');
          expect(labels.length, equals(2));
          expect(labels[0], equals('Item 1'));
          expect(labels[1], equals('Item 2'));
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  // Error handling tests

  group('LuaRunner Undefined Variable', () {
    test('throws on undefined variable access', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          expect(
            () => lua.run('print(undefined_variable.field)'),
            throwsA(isA<LuaException>()),
          );
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Create Invalid Collection', () {
    test('throws on nonexistent collection in script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          lua.run('db:create_element("Configuration", { label = "Test Config" })');
          expect(
            () => lua.run('db:create_element("NonexistentCollection", { label = "Item" })'),
            throwsA(isA<LuaException>()),
          );
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Empty Script', () {
    test('runs empty script successfully', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          // Empty script should succeed without error
          lua.run('');
          // If we get here, the test passed
          expect(true, isTrue);
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Comment Only Script', () {
    test('runs comment-only script successfully', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          // Comment-only script should succeed
          lua.run('-- this is just a comment');
          expect(true, isTrue);
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Read Integers', () {
    test('reads scalar integers in Lua script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        db.createElement('Configuration', {'label': 'Config'});
        db.createElement('Collection', {'label': 'Item 1', 'some_integer': 100});
        db.createElement('Collection', {'label': 'Item 2', 'some_integer': 200});

        final lua = LuaRunner(db);
        try {
          lua.run('''
            local ints = db:read_scalar_integers("Collection", "some_integer")
            assert(#ints == 2, "Expected 2 integers")
            assert(ints[1] == 100, "First integer mismatch")
            assert(ints[2] == 200, "Second integer mismatch")
          ''');
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Read Doubles', () {
    test('reads scalar doubles in Lua script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        db.createElement('Configuration', {'label': 'Config'});
        db.createElement('Collection', {'label': 'Item 1', 'some_float': 1.5});
        db.createElement('Collection', {'label': 'Item 2', 'some_float': 2.5});

        final lua = LuaRunner(db);
        try {
          lua.run('''
            local floats = db:read_scalar_doubles("Collection", "some_float")
            assert(#floats == 2, "Expected 2 doubles")
            assert(floats[1] == 1.5, "First double mismatch")
            assert(floats[2] == 2.5, "Second double mismatch")
          ''');
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Read Vectors', () {
    test('reads vector integers in Lua script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        db.createElement('Configuration', {'label': 'Config'});
        db.createElement('Collection', {'label': 'Item 1', 'value_int': [1, 2, 3]});

        final lua = LuaRunner(db);
        try {
          lua.run('''
            local vectors = db:read_vector_integers("Collection", "value_int")
            assert(#vectors == 1, "Expected 1 vector")
            assert(#vectors[1] == 3, "Expected 3 elements in vector")
            assert(vectors[1][1] == 1, "First element mismatch")
            assert(vectors[1][2] == 2, "Second element mismatch")
            assert(vectors[1][3] == 3, "Third element mismatch")
          ''');
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });

  group('LuaRunner Create With Vector', () {
    test('creates element with vector in Lua script', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final lua = LuaRunner(db);
        try {
          lua.run('''
            db:create_element("Configuration", { label = "Config" })
            db:create_element("Collection", { label = "Item 1", value_int = {10, 20, 30} })
          ''');

          final result = db.readVectorIntegers('Collection', 'value_int');
          expect(result.length, equals(1));
          expect(result[0], equals([10, 20, 30]));
        } finally {
          lua.dispose();
        }
      } finally {
        db.close();
      }
    });
  });
}

File: bindings/js/test/lua-runner.test.ts
import { describe, expect, test } from "bun:test";

const __dirname = import.meta.dir;

import { join } from "node:path";
import { Database, LuaRunner, QuiverError } from "../src/index.ts";

const SCHEMA_PATH = join(__dirname, "..", "..", "..", "tests", "schemas", "valid", "all_types.sql");

describe("LuaRunner", () => {
  test("create element from Lua and verify via JS", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      runner.run('db:create_element("AllTypes", { label = "FromLua" })');
      const labels = db.readScalarStrings("AllTypes", "label");
      expect(labels.includes("FromLua")).toBeTruthy();
    } finally {
      runner.close();
      db.close();
    }
  });

  test("Lua syntax error throws QuiverError", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      expect(() => runner.run("if then")).toThrow(QuiverError);
    } finally {
      runner.close();
      db.close();
    }
  });

  test("Lua runtime error throws QuiverError", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      expect(() => runner.run("local x = nil; x.field = 1")).toThrow(QuiverError);
    } finally {
      runner.close();
      db.close();
    }
  });

  test("multiple run calls on same runner succeed", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      runner.run('db:create_element("AllTypes", { label = "First" })');
      runner.run('db:create_element("AllTypes", { label = "Second" })');
      const labels = db.readScalarStrings("AllTypes", "label");
      expect(labels.length).toEqual(2);
    } finally {
      runner.close();
      db.close();
    }
  });

  test("empty script succeeds", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      // If this throws, the test fails automatically
      runner.run("");
    } finally {
      runner.close();
      db.close();
    }
  });

  test("close is idempotent", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      runner.close();
      // Second close should not throw
      runner.close();
    } finally {
      db.close();
    }
  });

  test("run after close throws QuiverError", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      runner.close();
      expect(() => runner.run("print('hello')")).toThrow(QuiverError);
    } finally {
      db.close();
    }
  });
});

describe("LuaRunner return values", () => {
  test("returns the script's value as JSON", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      expect(runner.run("return { a = 1, b = { 2, 3 } }")).toBe('{"a":1,"b":[2,3]}');
      expect(JSON.parse(runner.run("return db:read_element_ids('AllTypes')"))).toEqual([]);
    } finally {
      runner.close();
      db.close();
    }
  });

  test("returns an empty string when the script returns nothing", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      expect(runner.run("local x = 1")).toBe("");
    } finally {
      runner.close();
      db.close();
    }
  });
});

describe("Database dry run", () => {
  test("rolls back a script's writes", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    const runner = new LuaRunner(db);
    try {
      expect(db.inDryRun()).toBe(false);
      db.beginDryRun();
      expect(db.inDryRun()).toBe(true);

      // db:transaction composes: the dry run absorbs the nested BEGIN/COMMIT.
      const result = runner.run(`
        db:transaction(function(db)
          db:create_element("AllTypes", { label = "Preview" })
        end)
        return db:read_scalar_strings("AllTypes", "label")
      `);
      expect(JSON.parse(result)).toEqual(["Preview"]);

      db.endDryRun();
      expect(db.inDryRun()).toBe(false);
      expect(db.readScalarStrings("AllTypes", "label")).toEqual([]);
    } finally {
      runner.close();
      db.close();
    }
  });

  test("endDryRun without a dry run throws", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      expect(() => db.endDryRun()).toThrow(QuiverError);
    } finally {
      db.close();
    }
  });
});
